lib/hsa/activity_prof.h

```diff
@@ -0,0 +1,179 @@
+/*
+Copyright (c) 2015-2016 Advanced Micro Devices, Inc. All rights reserved.
+Permission is hereby granted, free of charge, to any person obtaining a copy
+of this software and associated documentation files (the "Software"), to deal
+in the Software without restriction, including without limitation the rights
+to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+copies of the Software, and to permit persons to whom the Software is
+furnished to do so, subject to the following conditions:
+The above copyright notice and this permission notice shall be included in
+all copies or substantial portions of the Software.
+THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANNTY OF ANY KIND, EXPRESS OR
+IMPLIED, INNCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+FITNNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT.  IN NO EVENT SHALL THE
+AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANNY CLAIM, DAMAGES OR OTHER
+LIABILITY, WHETHER INN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+OUT OF OR INN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+THE SOFTWARE.
+*/
+
+#ifndef ACTIVITY_PROF_H
+#define ACTIVITY_PROF_H
+
+#include <mutex>
+
+#include "kalmar_runtime.h"
+
+#if USE_PROF_API
+#include "prof_protocol.h"
+#include "hsa_rt_utils.hpp"
+
+#define ACTIVITY_PROF_INSTANCES()                                                    \
+    namespace activity_prof {                                                        \
+        CallbacksTable::table_t CallbacksTable::_table{};                            \
+        CallbacksTable::mutex_t CallbacksTable::_mutex;                              \
+        ActivityProf::timer_t* ActivityProf::_timer = NULL;                          \
+        std::atomic<record_id_t> ActivityProf::_glob_record_id(0);                   \
+    } // activity_prof
+
+namespace activity_prof {
+typedef activity_correlation_id_t record_id_t;
+typedef activity_op_t op_id_t;
+typedef Kalmar::hcCommandKind command_id_t;
+
+typedef activity_id_callback_t id_callback_fun_t;
+typedef activity_async_callback_t callback_fun_t;
+typedef void* callback_arg_t;
+
+class CallbacksTable {
+    public:
+    struct table_t {
+        id_callback_fun_t id_callback;
+        callback_fun_t fun[hc::HSA_OP_ID_NUM];
+        callback_arg_t arg[hc::HSA_OP_ID_NUM];
+    };
+    typedef std::recursive_mutex mutex_t;
```

Comment:
Don't think the recursive_mutex is needed anymore after the recursive call in set_async_callback has been removed, could you change this into a std::mutex?

Reply:
yes, will fix it, thanks